File: clients/typescript/e2e/e2e.ts
// One end to end run against a live core.
//
//     node e2e/e2e.ts http://127.0.0.1:8080 TOKEN
//
// Connect, subscribe, receive the snapshot and the flush, add a test source,
// take it, see `event/program.took`, disconnect. Every step prints ok, or the
// script exits non zero saying what failed.
//
// `clients/typescript/e2e/run.sh` starts a core the way dev/smoke.sh does and
// runs this against it.

import { connect, UI_EVENTS } from "../src/index.ts";

const base = process.argv[2] || "http://127.0.0.1:8080";
const token = process.argv[3] || null;
const ID = "e2e-typescript";

let failed = 0;

async function step<T>(name: string, work: () => Promise<T>): Promise<T> {
  process.stdout.write(name.padEnd(46));
  try {
    const value = await work();
    console.log("ok");
    return value;
  } catch (e) {
    console.log("FAIL");
    console.error("    " + (e instanceof Error ? e.message : String(e)));
    failed += 1;
    process.exit(1);
  }
}

const client = await step("connect to /rpc", () =>
  connect({ base, token, events: UI_EVENTS, autoSubscribe: false }),
);

const subscribed = await step("core.subscribe", () =>
  client.coreSubscribe({ events: UI_EVENTS, ext: { tally: true } }),
);
if (subscribed.ignored_ext.length) {
  console.log(`    this core ignored ext keys: ${subscribed.ignored_ext.join(", ")}`);
}

const state = await step("event/snapshot then event/flush", () => client.settled(10000));
console.log(`    flushed at seq ${state.seq}, ${state.sources.length} sources`);

const took = new Promise<string>((resolve, reject) => {
  const timer = setTimeout(() => reject(new Error("no event/program.took in ten seconds")), 10000);
  const off = client.on("event", (e) => {
    if (e.name !== "program.took") return;
    const params = e.params as { source?: string | null };
    if (params.source !== ID) return;
    clearTimeout(timer);
    off();
    resolve(params.source);
  });
});

const added = await step("source.add test://smpte", () =>
  client.sourceAdd({ id: ID, uri: "test://smpte", name: "TypeScript end to end" }),
);

await step("program.take", () => client.programTake({ source: added.id }));
const onAir = await step("event/program.took", () => took);
console.log(`    programme is ${onAir}`);

await step("source.remove", () => client.sourceRemove({ id: added.id }));

client.close();
console.log("disconnect".padEnd(46) + "ok");
process.exit(failed ? 1 : 0);

File: clients/python/e2e/e2e.py
#!/usr/bin/env python3
"""One end to end run against a live core.

    python3 clients/python/e2e/e2e.py http://127.0.0.1:8080 TOKEN

Connect, subscribe, receive the snapshot and the flush, add a test source, take
it, see `event/program.took`, disconnect. Every step prints ok, or the script
exits non zero saying what failed.

`clients/python/e2e/run.sh` starts a core the way dev/smoke.sh does and runs
this against it.
"""

from __future__ import annotations

import asyncio
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import godwinmix  # noqa: E402

ID = "e2e-python"


def step(name: str) -> None:
    print(f"{name:<46}", end="", flush=True)


def ok(detail: str = "") -> None:
    print("ok")
    if detail:
        print(f"    {detail}")


def fail(why: str) -> None:
    print("FAIL")
    print(f"    {why}", file=sys.stderr)
    raise SystemExit(1)


async def main() -> None:
    base = sys.argv[1] if len(sys.argv) > 1 else os.environ.get("GMX_URL", "http://127.0.0.1:8080")
    token = sys.argv[2] if len(sys.argv) > 2 else os.environ.get("GMX_TOKEN")

    step("connect to /rpc")
    try:
        client = await godwinmix.connect(base, token)
    except OSError as e:
        fail(str(e))
    ok()

    took: asyncio.Future = asyncio.get_running_loop().create_future()

    def on_took(params):
        if params.get("source") == ID and not took.done():
            took.set_result(params["source"])

    client.on("program.took", on_took)

    step("core.subscribe")
    try:
        result = await client.subscribe(ext={"tally": True})
    except godwinmix.RpcError as e:
        fail(str(e))
    ok(f"ignored ext keys: {result['ignored_ext']}" if result.get("ignored_ext") else "")

    step("event/snapshot then event/flush")
    try:
        state = await client.settled(10)
    except godwinmix.RpcError as e:
        fail(str(e))
    ok(f"flushed at seq {state['seq']}, {len(state['sources'])} sources")

    step("source.add test://smpte")
    try:
        added = await client.source_add(uri="test://smpte", id=ID, name="Python end to end")
    except godwinmix.RpcError as e:
        fail(str(e))
    ok(f"id {added['id']}, state {added.get('state')}")

    step("program.take")
    try:
        program = await client.take(added["id"])
    except godwinmix.RpcError as e:
        fail(f"{e}\n    {e.next_step}")
    ok(f"programme is {program.get('program')}")

    step("event/program.took")
    try:
        await asyncio.wait_for(took, 10)
    except asyncio.TimeoutError:
        fail(f"no event/program.took naming {ID} in ten seconds")
    ok()

    step("a picture, by whichever route this core has")
    try:
        jpeg = next(godwinmix.preview_stream(client, "program", width=320))
        if jpeg[:2] != b"\xff\xd8":
            fail(f"what came back is not a JPEG: {jpeg[:16]!r}")
    except (OSError, StopIteration) as e:
        fail(f"no picture: {e}")
    ok(f"{len(jpeg)} bytes")

    step("source.remove")
    try:
        await client.source_remove(id=added["id"])
    except godwinmix.RpcError as e:
        fail(str(e))
    ok()

    step("disconnect")
    await client.close()
    ok()


if __name__ == "__main__":
    asyncio.run(main())
